readme.md
# ASCII Display Demo

This repository contains the design and implementation of a command line script that displays an [ASCII Art](https://en.wikipedia.org/wiki/ASCII_art) text file in different colours. This calls back to when I used to mess around with batch files in senior high school. The only one I still have to this day is a script that displays a heart ASCII picture in different colours. As you might expect, it was very messy and bloated. I had no idea what I was doing back then.

As a quick experiment, I decided to recreate that old script using the programming techniques I know today. I also made a [flowchart](./pseudocode/readme.md) of the algorithm so that it can be implemented in other programming languages. Long story short, Windows batch scripting can be very powerful in the right hands when it comes to system administration. However, it lacks a lot of the complexities in today's programming languages that we take for granted.

---

## Getting Started

* To run the Windows batch script, execute `windows.bat`
* To run the Node JS script, execute `javascript.js`
* To run the Python script, execute `python.py`

---

## Credits

* [@tjohnston-softdev](https://github.com/tjohnston-softdev) 
	* Script design.
	* Windows Batch file implementation.
	* Node JS implementation.
	* Python implementation.
* Pseudocode Flowchart
	* Created using [Draw.IO](https://app.diagrams.net/)
	* Symbols adapted from [Rice University Fortran 90 Manual](https://www.owlnet.rice.edu/~ceng303/manuals/fortran/FOR3_3.html)
* Sample ASCII art from [ASCII Art Archive](https://www.asciiart.eu/computers/bug)
* ANSI colour codes from [Wikipedia](https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit)
* [Node JS text colouring tutorial](https://blog.bitsrc.io/coloring-your-terminal-using-nodejs-eb647d4af2a2) written by [@philipszdavido](https://github.com/philipszdavido)
	* Also applicable to Python.
* [Python clear terminal tutorial](https://www.tutorialspoint.com/how-to-clear-screen-in-python) written by [Pradeep Elance](https://www.tutorialspoint.com/answers/pradeep-elance) from [Tutorialspoint](https://www.tutorialspoint.com)

---

## Disclaimer
This project is licensed as [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/). You may use the contents for both personal and commercial purposes as long as credit is given, and any derived works are shared under the same license. This also includes any implementations of the design in different languages.

![CC BY-SA 4.0](https://i.creativecommons.org/l/by-sa/4.0/88x31.png)


changelog.md
# Changelog

Wrote ASCII display demo using Python

javascript.js
// ASCII Display Demo - Node JS

const fs = require("fs");
const inputFileName = "ascii.txt";
const maxFileSize = 100000;
const placeholderText = "ASCII art goes here.";
const supportedColours = [90, 91, 92, 93, 94, 95, 96, 97];
runAsciiDisplay();


// Main function.
function runAsciiDisplay()
{
	var inputFileSize = getInputSize();
	var retrievedContents = handleInputRead(inputFileSize);
	
	if (retrievedContents.length > 0)
	{
		callDisplayLoop(retrievedContents);
	}
}



// Retrieves ASCII input.
function getInputSize()
{
	var statObj = null;
	var retrievedSize = -1;
	
	try
	{
		// Open file and read size.
		statObj = fs.statSync(inputFileName);
		retrievedSize = statObj.size;
	}
	catch(fileError)
	{
		if (fileError.code === "ENOENT")
		{
			// File does not exist. Return empty.
			retrievedSize = -1;
		}
		else
		{
			// Error retrieving size.
			throw fileError;
		}
	}
	
	return retrievedSize;
}


// Reads input ASCII file.
function handleInputRead(inpSize)
{
	var readRes = "";
	
	if (inpSize > 0 && inpSize <= maxFileSize)
	{
		// Read input file.
		readRes = fs.readFileSync(inputFileName, "utf8");
	}
	else if (inpSize > maxFileSize)
	{
		// File too large.
		throw new Error("Input file cannot be larger than 100kb");
	}
	else
	{
		// Create input file.
		fs.writeFileSync(inputFileName, placeholderText, "utf8");
		readRes = placeholderText;
	}
	
	return readRes;
}


// ASCII display loop.
function callDisplayLoop(retInp)
{
	var colourValues = {current: -1, previous: 37};
	
	
	// Run every second.
	setInterval(function()
	{
		chooseNextColour(colourValues);
		displayAsciiArt(retInp, colourValues.current);
		setPreviousColour(colourValues);
	}, 1000);
}


// Displays ASCII art to console.
function displayAsciiArt(asciiText, colourValue)
{
	var setColour = "\x1b[" + colourValue + "m";
	var preparedText = [setColour, asciiText, "\x1b[39m"].join("");
	
	console.clear();
	console.log(preparedText);
}



// Choose next colour at random.
function chooseNextColour(colObj)
{
	var currentRandomIndex = -1;
	var currentValue = -1;
	var choiceMade = false;
	
	// Loop until a different colour has been chosen.
	while (choiceMade !== true)
	{
		currentRandomIndex = getRandomIndex();
		currentValue = supportedColours[currentRandomIndex];
		
		if (currentValue !== colObj.previous)
		{
			// Complete.
			colObj.current = currentValue;
			choiceMade = true;
		}
	}
}


// Save previous colour so it cannot be chosen next.
function setPreviousColour(colObj)
{
	var currentLocal = colObj.current;
	colObj.previous = currentLocal;
}



// Get random colour index.
function getRandomIndex()
{
	var seedValue = Math.random() * supportedColours.length;
	var randRes = Math.floor(seedValue);
	return randRes;
}



/*
	REFERENCES:
		(Title)								(URL)																					(Date Retrieved)
		Colour Codes						https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit							2021-04-06
		Display ASCII art to console.		https://blog.bitsrc.io/coloring-your-terminal-using-nodejs-eb647d4af2a2					2021-04-06
*/

src-nodejs/inp-file.js
const fs = require("fs");


// Check input file - Main.
function checkFileEntry(tgtPath)
{
	var dataObject = {};
	var checkResult = {flag: -1, message: ""};
	
	try
	{
		// Read and validate file entry.
		dataObject = fs.statSync(tgtPath);
		validateFileData(dataObject, checkResult);
	}
	catch(e)
	{
		// Error retrieving file entry.
		handleStatError(e.code, checkResult);
	}
	
	return checkResult;
}


// Read input file contents - Main.
function readFileContents(tgtPath)
{
	var readResult = defineAsciiText();
	var flaggedMessage = "";
	
	try
	{
		// Open and read.
		readResult.asciiText = fs.readFileSync(tgtPath, "utf8");
	}
	catch(e)
	{
		// Error.
		readResult = null;
		flaggedMessage = writeFileErrorMessage("reading", e.code);
		throw new Error(flaggedMessage);
	}
	
	return readResult;
}


// Create placeholder input file - Main.
function createPlaceholderFile(tgtPath)
{
	var createResult = defineAsciiText();
	var flaggedMessage = "";
	
	try
	{
		// Create placeholder.
		fs.writeFileSync(tgtPath, placeholderText, "utf8");
		createResult.asciiText = placeholderText;
	}
	catch(e)
	{
		// Error.
		createResult = null;
		flaggedMessage = writeFileErrorMessage("creating", e.code);
		throw new Error(flaggedMessage);
	}
	
	return createResult;
}


// Validate input file entry.
function validateFileData(dataObj, resObj)
{
	var correctType = dataObj.isFile();
	
	if (correctType === true && dataObj.size > 0 && dataObj.size <= maxSize)
	{
		// Valid.
		resObj.flag = 1;
	}
	else if (correctType === true && dataObj.size > maxSize)
	{
		// Too large.
		resObj.flag = -1;
		resObj.message = "Input file cannot be larger than 100kb";
	}
	else if (correctType === true)
	{
		// Empty - Create placeholder.
		resObj.flag = 0;
	}
	else
	{
		// Invalid file.
		resObj.flag = -1;
		resObj.message = "Input must be a valid file.";
	}
}


// Missing file entry.
function handleStatError(errorCodeString, resObj)
{
	if (errorCodeString === "ENOENT")
	{
		// File does not exist - Create placeholder.
		resObj.flag = 0;
	}
	else
	{
		// Error.
		resObj.flag = -1;
		resObj.message = writeFileErrorMessage("checking", errorCodeString);
	}
}



// Writes error message.
function writeFileErrorMessage(vAction, vErrorCode)
{
	var writeRes = "";
	
	writeRes += "Error ";
	writeRes += vAction;
	writeRes += " input file. - ";
	writeRes += parseErrorCode(vErrorCode);
	
	return writeRes;
}



// Converts FS error code to readable description.
function parseErrorCode(eCode)
{
	var parsedText = "";
	
	if (eCode === "EACCES")
	{
		parsedText = "Operation forbidden by access permissions.";
	}
	else if (eCode === "EISDIR")
	{
		parsedText = "Input path refers to a directory.";
	}
	else if (eCode === "EMFILE")
	{
		parsedText = "Too many files are open at once.";
	}
	else if (eCode === "ENOENT")
	{
		parsedText = "File does not exist.";
	}
	else if (eCode === "EPERM")
	{
		parsedText = "Elevated permissions required.";
	}
	else
	{
		// Other
		parsedText = "Unknown cause.";
	}
	
	return parsedText;
}



// ASCII text object.
function defineAsciiText()
{
	var defineRes = {asciiText: ""};
	return defineRes;
}




module.exports =
{
	checkEntry: checkFileEntry,
	readContents: readFileContents,
	createPlaceholder: createPlaceholderFile
};

python.py
# ASCII Display Demo - Python

import os
import time
import random

# Reads size of input file.
def getInputFileSize():
  global inputFileName
  fileExists = os.path.exists(inputFileName)
  sizeRes = -1
  
  if fileExists:
    # Get file size in bytes.
    sizeRes = os.path.getsize(inputFileName)
  else:
    # Return empty.
    sizeRes = -1
  
  return sizeRes
# End 'getInputFileSize'


# Creates or reads input file as needed.
def readInputContents():
  global inputFileName
  global entrySize
  fileEntry = None
  readRes = ""
  
  if (entrySize > 0 and entrySize <= maxFileSize):
    # Valid size - Read contents.
    fileEntry = open(inputFileName, "r")
    readRes = fileEntry.read()
    fileEntry.close()
  elif (entrySize > maxFileSize):
    # File too large.
    raise Exception("Input file cannot be larger than 100kb")
  else:
    # Empty or missing - Create placeholder.
    fileEntry = open(inputFileName, "w")
    fileEntry.write(placeholderText)
    fileEntry.close()
    readRes = placeholderText
  
  return readRes
# End 'readInputContents'


# Chooses next colour for ASCII print.
def chooseRandomColour():
  global colourObject
  currentRandomIndex = -1
  currentValue = -1
  choiceMade = False
  
  # Update previous colour.
  previousColour = colourObject["current"]
  colourObject["previous"] = colourObject["current"]
  
  # Loop until different colour is chosen.
  while (choiceMade != True):
    
    # Choose random colour.
    currentRandomIndex = random.randrange(0, len(supportedColours))
    currentValue = supportedColours[currentRandomIndex]
    
    if (currentValue != previousColour):
      # Update current colour.
      colourObject["current"] = currentValue
      choiceMade = True
    # End If
  # End While
# End 'chooseRandomColour'


# Double checks that retrieved ASCII is not empty.
def checkOutputReady():
  global retrievedText
  strLength = len(retrievedText)
  checkRes = bool(strLength > 0)
  return checkRes
# End 'checkOutputReady'

# Clears console.
def clearConsole():
  if (os.name == "posix"):
    os.system("clear")
  else:
    os.system("cls")
  # End If
# End 'clearConsole'


# Displays ASCII text to console in current colour.
def displayAsciiArt():
  global retrievedText
  global colourObject
  
  colourValue = colourObject["current"]
  setColour = "\x1b[" + str(colourValue) + "m"
  preparedText = setColour + retrievedText + "\x1b[39m"
  print(preparedText)
# End 'displayAsciiArt'


# Waits one second before next print.
def handleWait():
  waitSuccessful = False
  
  # Using 'try-except' avoids error on terminate.
  try:
    time.sleep(1)
    waitSuccessful = True
  except:
    waitSuccessful = False
  # End 'try-except'
  
  return waitSuccessful
# End 'handleWait'


# Begin main code:

supportedColours = (90, 91, 92, 93, 94, 95, 96, 97)
inputFileName = "ascii.txt"
maxFileSize = 100000
placeholderText = "ASCII art goes here."
entrySize = getInputFileSize()
retrievedText = readInputContents()
colourObject = {"current": -1, "previous": 37}
canContinue = checkOutputReady()


# Loop forever.
while (canContinue == True):
  chooseRandomColour()
  clearConsole()
  displayAsciiArt()
  canContinue = handleWait()
# End while




# Same references as JavaScript.